native/src/widget/tooltip.rs

```diff
@@ -0,0 +1,239 @@
+pub struct Tooltip<'a, Message, Renderer: self::Renderer> {
+    state: State,
+    content: Element<'a, Message, Renderer>,
+    hover_content: Option<Element<'a, Message, Renderer>>,
+}
+
+impl<'a, Message, Renderer> Tooltip<'a, Message, Renderer>
+where
+    Renderer: self::Renderer,
+{
+    /// Creates an empty [`Tooltip`].
+    ///
+    /// [`Tooltip`]: struct.Tooltip.html
+    pub fn new<T>(content: T) -> Self
+    where
+        T: Into<Element<'a, Message, Renderer>>,
+    {
+        Tooltip {
+            state: Default::default(),
+            content: content.into(),
+            hover_content: None,
+        }
+    }
+
+    /// Sets the hover_content of the [`Tooltip`].
+    ///
+    /// [`Tooltip`]: struct.Tooltip.html
+    pub fn hover_content<T>(mut self, hover_content: T) -> Self
+    where
+        T: Into<Element<'a, Message, Renderer>>,
+    {
+        self.hover_content = Some(hover_content.into());
+        self
+    }
+}
+
+#[derive(Debug, Default)]
+struct State {
+    cursor_position: Point,
+    is_hovered: bool,
+}
+
+impl<'a, Message, Renderer> Widget<Message, Renderer>
+    for Tooltip<'a, Message, Renderer>
+where
+    Renderer: self::Renderer,
+{
+    fn width(&self) -> Length {
+        self.content.width()
+    }
+
+    fn height(&self) -> Length {
+        self.content.height()
+    }
+
+    fn layout(
+        &self,
+        renderer: &Renderer,
+        limits: &layout::Limits,
+    ) -> layout::Node {
+        self.content.layout(renderer, limits)
+    }
+
+    fn on_event(
+        &mut self,
+        event: Event,
+        layout: Layout<'_>,
+        cursor_position: Point,
+        messages: &mut Vec<Message>,
+        renderer: &Renderer,
+        clipboard: Option<&dyn Clipboard>,
+    ) {
+        match event {
+            Event::Mouse(iced_core::mouse::Event::CursorMoved { x, y }) => {
+                self.state.cursor_position.x = x;
+                self.state.cursor_position.y = y;
+                if layout.bounds().contains(self.state.cursor_position) {
+                    self.state.is_hovered = true;
+                } else {
+                    self.state.is_hovered = false;
+                }
+            }
+            _ => (),
+        }
+        self.content.widget.on_event(
+            event,
+            layout,
+            cursor_position,
+            messages,
+            renderer,
+            clipboard,
+        )
+    }
+
+    fn draw(
+        &self,
+        renderer: &mut Renderer,
+        defaults: &Renderer::Defaults,
+        layout: Layout<'_>,
+        cursor_position: Point,
+    ) -> Renderer::Output {
+        renderer.draw(defaults, cursor_position, &self.content, layout)
+    }
+
+    fn hash_layout(&self, state: &mut Hasher) {
+        struct Marker;
+        std::any::TypeId::of::<Marker>().hash(state);
+
+        self.content.hash_layout(state);
+    }
+
+    fn overlay(
+        &mut self,
+        layout: Layout<'_>,
```

Comment:
We could consider extending `Widget::overlay` to also provide a `cursor_position`. This should allow us to remove the local `State` of the `Tooltip`.

I'll think about it!

Reply:
Did that in b8c5ebca04d90c99a3fd887a525aca399e575613 and added a `overlay_content_bounds` field. I need it to calculate where to put the tooltip (if `tooltip_position` is different than `TooltipPosition::FollowCursor` in `Tooltip`). There is probably a better way to do it (?), but that's what I do currently.